### ISSUE TITLE
Can not move subfolders and their content between shared folders when using encryption and nextcloud desktop app (windows)

### ISSUE THREAD

**Issue by Kuphi** [nextcloud/server#24219]
This is a more complex one:

### Steps to reproduce
1. Use Nextcloud with encryption (user keys, master key and recovery key)
2. See picture
![Nextcloud error](https://user-images.githubusercontent.com/64360528/99679079-d3f00a80-2a7b-11eb-8a87-e131aa08af65.png)

### Expected behaviour
Subfolders and files should also be synchronized.

### Actual behaviour
Folder I, K and File L will never be synchronized

### Workaround
Crop and paste "Folder H" to any folder on your computer outside of the nextcloud. Then crop "Folder H" again and paste into your target nextcloud folder. I think previous file versions will/might get lost by using the workaround.

### Server configuration

**Operating system:** Ubuntu server 20.04.1
**Web server:** Apache2
**Database:** mySQL / mariaDB
**PHP version:** 7.4.3
**Nextcloud version:** (see Nextcloud admin page) 19.0.4
**Updated from an older Nextcloud/ownCloud or fresh install:** 18
**Where did you install Nextcloud from:** https://www.linuxbabe.com/ubuntu/install-nextcloud-ubuntu-20-04-apache-lamp-stack

### Log
{"reqId":"xxxxxxxxxxxxxxxxxxx","level":4,"time":"2020-11-19T13:51:32+00:00","remoteAddr":"123.456.789.123","user":"userB","app":"webdav","method":"MOVE","url":"/remote.php/dav/files/userB/folder1/Objekte/ARCHIV/folderH/folderI","message":{"Exception":"OCP\\Files\\NotFoundException","Message":"/server/files/folder2/Objekte/folderH/folderI/fileL","Code":0,"Trace":[{"file":"/var/www/nextcloud/lib/private/Files/Node/Folder.php","line":138,"function":"get","class":"OC\\Files\\Node\\Root","type":"->"},{"file":"/var/www/nextcloud/lib/private/Encryption/File.php","line":93,"function":"get","class":"OC\\Files\\Node\\Folder","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Stream/Encryption.php","line":266,"function":"getAccessList","class":"OC\\Encryption\\File","type":"->"},{"function":"stream_open","class":"OC\\Files\\Stream\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Stream/Encryption.php","line":207,"function":"fopen"},{"file":"/var/www/nextcloud/lib/private/Files/Stream/Encryption.php","line":187,"function":"wrapSource","class":"OC\\Files\\Stream\\Encryption","type":"::"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":473,"function":"wrap","class":"OC\\Files\\Stream\\Encryption","type":"::"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":756,"function":"fopen","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":653,"function":"copyBetweenStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":749,"function":"copyFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":653,"function":"copyBetweenStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":749,"function":"copyFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":625,"function":"copyBetweenStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php","line":575,"function":"moveFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php","line":575,"function":"moveFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Jail.php","line":514,"function":"moveFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php","line":575,"function":"moveFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Jail","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/View.php","line":822,"function":"moveFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->"},{"file":"/var/www/nextcloud/apps/dav/lib/Connector/Sabre/Directory.php","line":437,"function":"rename","class":"OC\\Files\\View","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Tree.php","line":164,"function":"moveInto","class":"OCA\\DAV\\Connector\\Sabre\\Directory","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php","line":641,"function":"move","class":"Sabre\\DAV\\Tree","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php","line":89,"function":"httpMove","class":"Sabre\\DAV\\CorePlugin","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":474,"function":"emit","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":251,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":319,"function":"start","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/apps/dav/lib/Server.php","line":320,"function":"exec","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/apps/dav/appinfo/v2/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/remote.php","line":167,"args":["/var/www/nextcloud/apps/dav/appinfo/v2/remote.php"],"function":"require_once"}],"File":"/var/www/nextcloud/lib/private/Files/Node/Root.php","Line":205,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Windows) mirall/3.0.3stable-Win64 (build 20201029) (Nextcloud)","version":"19.0.4.2","id":"xxxxxxxxxx"}

{"reqId":"xxxxxxxxxxxxxxxxxxxx","level":3,"time":"2020-11-19T13:51:32+00:00","remoteAddr":"123.456.789.123","user":"enrico","app":"PHP","method":"MOVE","url":"/remote.php/dav/files/enrico/folder1/Objekte/ARCHIV/folderH/folderI","message":{"Exception":"Error","Message":"fclose() expects parameter 1 to be resource, string given at /var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php#762","Code":0,"Trace":[{"function":"onError","class":"OC\\Log\\ErrorHandler","type":"::"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":762,"function":"fclose"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":653,"function":"copyBetweenStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":749,"function":"copyFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":653,"function":"copyBetweenStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":749,"function":"copyFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":625,"function":"copyBetweenStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php","line":575,"function":"moveFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php","line":575,"function":"moveFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Jail.php","line":514,"function":"moveFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php","line":575,"function":"moveFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Jail","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/View.php","line":822,"function":"moveFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->"},{"file":"/var/www/nextcloud/apps/dav/lib/Connector/Sabre/Directory.php","line":437,"function":"rename","class":"OC\\Files\\View","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Tree.php","line":164,"function":"moveInto","class":"OCA\\DAV\\Connector\\Sabre\\Directory","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php","line":641,"function":"move","class":"Sabre\\DAV\\Tree","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php","line":89,"function":"httpMove","class":"Sabre\\DAV\\CorePlugin","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":474,"function":"emit","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":251,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":319,"function":"start","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/apps/dav/lib/Server.php","line":320,"function":"exec","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/apps/dav/appinfo/v2/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type":"->"},{"file":"/var/www/nextcloud/remote.php","line":167,"args":["/var/www/nextcloud/apps/dav/appinfo/v2/remote.php"],"function":"require_once"}],"File":"/var/www/nextcloud/lib/private/Log/ErrorHandler.php","Line":91,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Windows) mirall/3.0.3stable-Win64 (build 20201029) (Nextcloud)","version":"19.0.4.2","id":"xxxxxxxxx"}

**Comment by Kuphi** [comment]
Can anyone confirm this bug? I think it's serious as it messes up the Nextloud a lot and may require restoring from a backup.

**Comment by Kuphi** [comment]
Already a month passed and no comment at all. This is quite a serious issue.

I can not advise all users to cut and paste the data to the desktop and then cut and paste again to the destination folder.

Thank you for improving nextcloud :)

**Comment by Kuphi** [comment]
Is there no answer, because the error is occurring at the client side?
I created a new issue for nextcloud desktop client: https://github.com/nextcloud/desktop/issues/2797

**Comment by Kuphi** [comment]
This issue has been fixed in the desktop branch: https://github.com/nextcloud/desktop/pull/2824  https://github.com/nextcloud/desktop/issues/2797
Thank you!

**Comment by jospoortvliet** [comment]
Hi @Kuphi you should try to do the same operation in the web UI. You will likely get an error - which is why the desktop client can not do this either.  You can't put a shared folder into another shared folder - that would lead to all kinds of weird situations. For example, what if User A then un-shares folder G with you. But folder F was IN folder G and it is still shared with you, so do you lose access to the now-subfolder of folder G or not? To avoid such complications, Nextcloud will simply not let you do this - it gives an error in the web interface and does not sync the files if you do it in the file manager. We can't STOP you from moving the files in the client, unfortunately, but the client does tell you that this was not synced as it gave an error.

When you move the folder out, you effectively delete the share and when you move the files back, they are no longer the original shared folder but new files. If user A now makes changes, those will NOT be reflected in your shared folder anymore.

In other words, this is not a bug but expected behavior.

**Comment by Kuphi** [comment]
@jospoortvliet Thank you for your reply! The issue has already been resolved by nextcloud desktop team.

You probably misunderstood me. I never moved a shared folder to a shared folder. I was just moving a folder from a shared folder to another folder. This worked only partially. This feature should surely work. It was a bug, now fixed.
